### PR TITLE
add rspecs for controllers and models

### PR DESCRIPTION
**Pull request by SerhiiBorozenets**
https://github.com/SerhiiBorozenets/react-project/issues/1

### CODE REVIEW

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    it { expect(subject).to have_many(:projects).dependent(:destroy) }
+  end
+
+  context 'Validations' do
+    it { is_expected.to validate_presence_of(:email) }
+    it { is_expected.to validate_presence_of(:password) }
+
+    it 'fails from 1 to 5 characters' do
+      user.password = Faker::Internet.password(min_length: 1, max_length: 5)
+      expect(user.valid?).to be_falsey
```
Comment (rogergraves) [inline spec/models/user_spec.rb:17]: Also can use `expect(user).to_not be_valid`

**Review by rogergraves — CHANGES_REQUESTED** [review]
Nice work for a first try on this 👍🏻

File: spec/controllers/api/v1/projects_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+      subject
+      expect(response.media_type).to eq "application/json"
+    end
+  end
+
+  describe "POST" do
+    it "creates a Project with data" do
+      subject
+
+      project_saved = Project.find_by(title: project[:title])
```
[the remark this answers is not in the document]
Comment (SerhiiBorozenets) [reply spec/controllers/api/v1/projects_controller_spec.rb:32]: @rogergraves unfortunately `project` from line 6 has only params `{ attributes_for :project, user_id: user.id }` , so I should find `saved project` in db and check fields.

File: spec/controllers/api/v1/projects_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                                                         }
+                                                     },
+                                                   "included" => []
+                                                 }]
+    end
+  end
+
+  describe "PATCH" do
+    it "updates a Project and returns data" do
+      project = Project.create(title: Faker::Hobby.activity, user_id: user.id)
```
[the remark this answers is not in the document]
Comment (SerhiiBorozenets) [reply spec/controllers/api/v1/projects_controller_spec.rb:63]: thanks

File: app/models/user.rb [hunk 4]
```diff
@@ -4,6 +4,9 @@ class User < ApplicationRecord
          :omniauthable, omniauth_providers: %i[google_oauth2 github facebook]
 
   has_many :projects, dependent: :destroy
+  validates :email, presence: true
+  validates :password, presence: true, length: { minimum: 6, maximum: 128 }
```
[the remark this answers is not in the document]
Comment (SerhiiBorozenets) [reply app/models/user.rb:8]: thanks

File: spec/controllers/api/v1/projects_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,99 @@
+require 'rails_helper'
+require 'shared_contexts'
+
+RSpec.describe Api::V1::ProjectsController, :type => :controller do
+  let!(:user) { create :user }
+  let!(:project) { attributes_for :project, user_id: user.id }
```
[the remark this answers is not in the document]
Comment (SerhiiBorozenets) [reply spec/controllers/api/v1/projects_controller_spec.rb:6]: @rogergraves I used it like params for creating project, so I renamed it to `let!(:project_attr) { attributes_for :project, user_id: user.id }`. Thanks

File: spec/controllers/api/v1/projects_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                                                         }
+                                                     },
+                                                   "included" => []
+                                                 }]
+    end
+  end
+
+  describe "DELETE" do
+    it "removes project" do
+      project = Project.create(title: Faker::Hobby.activity, user_id: user.id)
```
[the remark this answers is not in the document]
Comment (SerhiiBorozenets) [reply spec/controllers/api/v1/projects_controller_spec.rb:95]: thanks

File: spec/factories.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
+FactoryBot.define do
+  factory(:user) do
+    email { Faker::Internet.email }
+    password { Faker::Internet.password(min_length: 6, max_length: 128) }
+  end
+
+  factory(:project) do
+    title { Faker::Hobby.activity }
+  end
```
[the remark this answers is not in the document]
Comment (SerhiiBorozenets) [reply spec/factories.rb:9]: thanks

File: spec/factories.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+  factory(:task) do
+    title { Faker::Hobby.activity }
+    due_date { Faker::Date.forward(days: 30).to_s }
+    status { %w[No Low Middle High].sample }
+    complexity { %w[None Elementary Intermediate Advanced Master].sample }
+    completed { Faker::Boolean.boolean }
```
[the remark this answers is not in the document]
Comment (SerhiiBorozenets) [reply spec/factories.rb:16]: thanks

File: spec/models/project_spec.rb [hunk 9]
```diff
@@ -0,0 +1,40 @@
+require 'rails_helper'
+
+RSpec.describe Project, :type => :model do
+  let!(:user) { create :user }
```
[the remark this answers is not in the document]
Comment (SerhiiBorozenets) [reply spec/models/project_spec.rb:4]: thanks

File: spec/models/project_spec.rb [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    end
+
+    it "is valid with a description" do
+      project = Project.new(title: Faker::Hobby.activity, user_id: user.id)
+      expect(project).to be_valid
+    end
+
+    it "is create projects with the same title and others slugs" do
+      project_1 = Project.create(title: "The project", user_id: user.id)
+      project_2 = Project.create(title: "The project", user_id: user.id)
```
[the remark this answers is not in the document]
Comment (SerhiiBorozenets) [reply spec/models/project_spec.rb:25]: thanks

File: spec/models/project_spec.rb [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+      expect(project_1.slug).to_not eq project_2.slug
+    end
+
+    it "is slug present" do
+      project = Project.create(title: Faker::Hobby.activity, user_id: user.id)
+      expect(project.slug).to eq project.title.parameterize
+    end
+
+    it "is slug not changed" do
+      project = Project.create(title: Faker::Hobby.activity, user_id: user.id)
```
[the remark this answers is not in the document]
Comment (SerhiiBorozenets) [reply spec/models/project_spec.rb:35]: thanks

**Comment by SerhiiBorozenets** [comment]
@rogergraves please take a look when you have time https://github.com/SerhiiBorozenets/react-project/pull/20/commits/e8831cf39a4e0771cd8b880c3deaf0d80322b326

**Review by rogergraves — APPROVED** [review]
Nice work! 👍🏻

**Pull request merged**